kubernetes-kit-starter/src/test/java/com/vaadin/kubernetes/starter/sessiontracker/SessionSerializerTest.java

```diff
@@ -380,14 +385,39 @@ void serialize_notSerializableException_notFallbackToPessimistic() {
 
         serializer.serialize(httpSession);
         await().atMost(100, MILLISECONDS).untilTrue(serializationStarted);
-        verify(connector).markSerializationStarted(clusterSID);
+        verify(connector).markSerializationStarted(clusterSID, timeToLive);
 
         await().atMost(500, MILLISECONDS).untilTrue(serializationCompleted);
         verify(connector, never()).sendSession(any());
         verify(connector).markSerializationComplete(clusterSID);
         locks.forEach(l -> verify(l, times(1)).lock());
     }
 
+    @Test
+    void serialize_runtimeException_pendingSessionKeyRemoved() {
+        Throwable throwable = new RuntimeException("error");
+        AtomicBoolean serializationStarted = new AtomicBoolean();
+        doAnswer(i -> serializationStarted.getAndSet(true)).when(connector)
+                .markSerializationStarted(clusterSID, timeToLive);
+
+        AtomicBoolean serializationFailed = new AtomicBoolean();
+        doAnswer(i -> serializationFailed.getAndSet(true)).when(connector)
+                .markSerializationFailed(eq(clusterSID), any());
+
+        doThrow(throwable).when(connector).sendSession(any());
+
+        vaadinSession.setLockTimestamps(10, 20);
+
+        serializer.serialize(httpSession);
+        await().atMost(100, MILLISECONDS).untilTrue(serializationStarted);
+        verify(connector).markSerializationStarted(clusterSID, timeToLive);
+
+        await().atMost(500, MILLISECONDS).untilTrue(serializationFailed);
+        assertThrows(RuntimeException.class,
+                () -> connector.sendSession(any()));
+        verify(connector).markSerializationFailed(eq(clusterSID), any());
```

Comment:
If we unwrap `CompletionException` in `SessionSerializer` the we can verify that the exact exception is given to the connector


```suggestion
        verify(connector).markSerializationFailed(eq(clusterSID), same(throwable));
```